Batch Index Settings Update Requests

joegallo opened the pull request:
Closes #79866

Mostly follows the same shape as the changes in #82159.

~WIP because this currently fails `./gradlew ':server:test' --tests "org.elasticsearch.indices.cluster.IndicesClusterStateServiceRandomUpdatesTests.testRandomClusterStateUpdates"` pretty reliably. 😬~

joegallo committed:
Extract a clusterTask to pass, rather than on-the-fly

joegallo committed:
Just use metadata() here

joegallo committed:
Move the reroute *slightly* later

joegallo committed:
Whitespace

joegallo committed:
Assign these in constructor argument order

joegallo committed:
Use a reroute-after-tasks-finished executor

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:
```diff
@@ -67,11 +67,26 @@ public MetadataUpdateSettingsService(
+                }
+            }
+            // reroute in case things change that require it (like number of replicas)
+            state = allocationService.reroute(state, "settings update");
```
let's skip this in case the batch was a NOOP and `currentState == state`

joegallo replied:
👍, be4de6c

original-brownbear reviewed (commented):
Looks good, just one quick comment for now. We do need to figured out how/why this breaks tests though.

joegallo committed:
This gets the tests passing again

It's the first thing the reroute call itself actually called, so we're
keeping a portion of a full reroute call here, but deferring the rest
of it until the end of the executor loop.

joegallo committed:
Rename this variable, s/closeIndices/closedIndices/g

joegallo committed:
Merge branch 'master' into batch-settings-update

joegallo committed:
Skip the reroute if the whole batch was a noop

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

original-brownbear reviewed (approved):
LGTM, I'd like David to have a look at this too with the reroute weirdness. Maybe he has a better idea on how to cleanly handle this.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:
```diff
@@ -105,149 +122,153 @@ public void updateSettings(final UpdateSettingsClusterStateUpdateRequest request
-                        .build();
+                // we need to tweak auto expand replicas in order to avoid tripping assertions in
+                // AllocationService.reroute(RoutingAllocation allocation) -- this is far from ideal
+                updatedState = allocationService.adaptAutoExpandReplicas(updatedState);
```
I think the problem is here:

https://github.com/elastic/elasticsearch/blob/6959203b5aadf9f07db183a9a40a773a775557b9/server/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java#L468-L478

In these tests we run each individual `ClusterStateUpdateTask` but with this change we moved the tidy-up `reroute()` call outside the individual tasks so it's not running.

joegallo replied:
Great catch! I think 92bc39d should do the trick then, yes?

joegallo committed:
Rework test code, remove adaptAutoExpandReplicas hack

joegallo committed:
Merge branch 'master' into batch-settings-update

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/82896.yaml

DaveCTurner reviewed (approved):
LGTM

joegallo commented:
@elasticmachine run elasticsearch-ci/part-1

joegallo commented:
@elasticmachine run elasticsearch-ci/part-1

joegallo commented:
@elasticmachine run elasticsearch-ci/part-1

joegallo commented:
@elasticmachine run elasticsearch-ci/part-1

joegallo commented:
I opened https://github.com/elastic/elasticsearch/issues/83097 for some of the failures I saw here, and that's been picked up, so I'm merging this one.
